File: sdk/python/rhoban/motors.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys, os, re, threading, time
from datetime import datetime
import rhoban.configurations as config
import rhoban.communication as com

"""
    Classe permettant de gérer et de dispatcher un ensemble de moteurs
"""
class Motors(threading.Thread):
    def __init__(self, connection):
        super(Motors, self).__init__()
        self.connection = connection
        self.configuration = None
        self.motors = {}
        self.idMotors = {}
        self.running = False

    def setConfig(self, configuration):
        self.configuration = configuration

        for name, servo in self.configuration.servos.items():
            self.motors[name] = Motor(servo.id, name, servo.iniAngle, servo.zeroAngle)
            self.idMotors[int(servo.id)] = self.motors[name]

    def __len__(self):
        return len(self.motors)

    def __getitem__(self, key):
        return self.motors[key]

    def __iter__(self):
        return iter(self.motors)

    def start(self, frequency):
        self.frequency = frequency

        if self.configuration == None:
            raise Exception('No suitable motors configuration')

        self.running = True
        super(Motors, self).start()

    def stop(self):
        self.running = False

    def compilant(self, name):
        self.motors[name].setLoad(0)

    def allCompliant(self):
        self.connection.ServosAllCompliant()

        for name, motor in self.motors.items():
            motor.setLoad(0)

    def hard(self, name):
        self.motors[name].setLoad(1)

    def allHard(self):
        for name, motor in self.motors.items():
            motor.setLoad(1)

    def pullValues(self):
        values = self.connection.ServosGetValues_response(1)
        self.processValues(values)

    def pushValues(self):
        ids = []
        angles = []
        speeds = []
        loads = []

        for name, motor in self.motors.items():
            if motor.goalAngle != None and motor.currentAngle != None and motor.currentSpeed != None and motor.dirty:
                ids += [motor.id]
                angles += [int(motor.goalAngle*1000)]
                speeds += [int(motor.goalSpeed*1023)]
                loads += [int(motor.goalLoad*1023)]

        if len(ids):
            self.connection.ServosSetValues(1, ids, angles, speeds, loads)
            
    def processValues(self, values):
        for id, angle, speed, load in zip(values[0], values[1], values[2], values[3]):
            if id in self.idMotors:
                motor = self.idMotors[int(id)]
                motor.lastUpdate = datetime.now()

                motor.currentAngle = angle
                if motor.goalAngle == None:
                    motor.goalAngle = motor.currentAngle

                motor.currentSpeed = speed/1023.0
                if motor.goalSpeed == None:
                    motor.goalSpeed = max(0, motor.currentSpeed)

                motor.currentLoad = load/1023.0
                if motor.goalLoad == None:
                    motor.goalLoad = max(0, motor.currentLoad)
    
    def goToZero(self, duration = 5, verbose = False):
        self.pullValues()
        for name, motor in self.motors.items():
            motor.setAngle(motor.zeroAngle)

        self.raiseLoad(duration, verbose)

    def goToInit(self, duration = 5, verbose = False):
        self.pullValues()
        for name, motor in self.motors.items():
            motor.setRelAngle(0)

        self.raiseLoad(duration, verbose)

    def raiseLoad(self, duration = 5, verbose = False):
        cs = duration * 100
        for x in range(cs):
            if verbose:
                sys.stdout.write("\rLoad: %3d%%" % (round(100*(x+1)/float(cs))))
                sys.stdout.flush()
            for name, motor in self.motors.items():
                motor.setLoad((x +1) / float(cs))
            time.sleep(0.01)
            self.pushValues()
        if verbose:
            print('')
            
    def run(self):
        motors = self

        def motorsValues(values):
            motors.processValues(values)

        while self.running:
            self.connection.ServosGetValues_callback(1, motorsValues)
            self.pushValues()
            time.sleep(1.0/self.frequency)

    def scan(self):
        for name, motor in self.motors.items():
            motor.lastUpdate = None

        self.connection.ServosScan(250, 'Normal')
"""
    Représente un moteur
"""
class Motor:
    def __init__(self, id, name, iniAngle, zeroAngle):
        self.lastUpdate = None
        self.id = int(id)
        self.name = name
        self.iniAngle = float(iniAngle)
        self.zeroAngle = float(zeroAngle)

        self.goalAngle = None
        self.currentAngle = None

        self.goalSpeed = None
        self.currentSpeed = None

        self.goalLoad = None
        self.currentLoad = None

        self.dirty = False

    def setAngle(self, angle):
        self.dirty = True
        self.goalAngle = float(angle)

    def setRelAngle(self, angle):
        self.dirty = True
        self.goalAngle = self.iniAngle + self.zeroAngle + float(angle)

    def setLoad(self, load):
        self.dirty = True
        self.goalLoad = float(load)

    def setSpeed(self, speed):
        self.dirty = True
        self.goalSpeed = float(speed)

    def getAngle(self):
        return float(self.currentAngle)

    def getRelAngle(self):
        return float(self.currentAngle - self.iniAngle - self.zeroAngle)

    def getSpeed(self):
        return float(self.currentSpeed)

    def getLoad(self):
        return float(self.currentLoad)

File: sdk/cpp/rhoban/motors/Motors.cpp
/*************************************************
 * Publicly released by Rhoban System, August 2012
 *             www.rhoban-system.fr
 *
 * Freely usable for non-commercial purposes
 *
 * Licence Creative Commons *CC BY-NC-SA
 * http://creativecommons.org/licenses/by-nc-sa/3.0
 *************************************************/
#include <cstdlib>
#include <cstdio>
#include <map>
#include <string>
#include <threading/Thread.h>
#include <config/MoveSchedulerConfig.h>
#include <communication/Connection.h>
#include <robot/Robot.h>
#include <types.h>
#include "Motor.h"
#include "Motors.h"
#include "ticks.h"

using namespace std;

namespace Rhoban
{
    void motorsValues(Message *values, void *data)
    {
        Motors *self = (Motors *)data;
        self->processValues(values);
    }

    Motors::Motors(Connection *connection) : Thread()
    {
        this->connection = connection;
        this->configuration = NULL;
        running = 0;
    }

    Motors::~Motors()
    {
    }

    void Motors::setConfig(MoveSchedulerConfig *config)
    {
        this->configuration = config;

        map<string, ServoConfig>::iterator it;
        for(it = configuration->servos_config.servos.begin();
                it != configuration->servos_config.servos.end(); ++it)
        {
            ServoConfig &config = (*it).second;
            string name = config.Name;
            motors[name] = new Motor(config.Id, config.Name, config.IniAngle, config.ZeroAngle);
            idMotors[config.Id] = motors[name];
        }
    }

    size_t Motors::len()
    {
        return motors.size();
    }

    Motor *Motors::operator[](string name)
    {
        return get(name);
    }

    Motor *Motors::get(string name)
    {
        if (motors.find(name) != motors.end()) {
            return motors[name];
        } else {
            ostringstream oss;
            oss << "Unable to find motor with name " << name;

            throw oss.str();
        }
    }

    void Motors::start(int frequency)
    {
        this->frequency = frequency;

        if(configuration == NULL)
            cout << "No suitable motors configuration" << endl;
        else
        {
            running = 1;
            Thread::start(NULL);
        }
    }

    void Motors::stop()
    {
        running = 0;
    }

    void Motors::compliant(string name)
    {
        motors[name]->setLoad(0);
    }

    void Motors::allCompliant()
    {
        connection->ServosAllCompliant();

        map<string, Motor *>::iterator it;
        for(it = motors.begin(); it != motors.end(); ++it)
            it->second->setLoad(0);
    }

    void Motors::hard(string name)
    {
        motors[name]->setLoad(1);
    }

    void Motors::allHard()
    {
        map<string, Motor *>::iterator it;
        for(it = motors.begin(); it != motors.end(); ++it)
            it->second->setLoad(1);
    }

    void Motors::pullValues()
    {
        try
        {
            Message values = connection->ServosGetValues_response(1,10000);
            processValues(&values);
        }
        catch(string exc)
        {
            throw string("Failed to pull values:\n\t") + exc;
        }
    }

    void Motors::pushValues()
    {
        map<string, Motor *>::iterator it;

        vector<float> angles;
        vector<string> names;

        for(it = motors.begin(); it != motors.end(); ++it) {
            if(it->second->isDirty()
                    && it->second->getGoalAngleInit()
                    && it->second->getCurrentAngleInit()
                    && it->second->getCurrentSpeedInit()) {
                names.push_back(it->first);
                angles.push_back((int)  (it->second->getGoalAngle()) );
            }
        }

        if(names.size()) {
            connection->SetRelativeServosAngles(names, angles);
        }
    }

    void Motors::processValues(Message *values)
    {
        vector<byte> ids = values->read_array();
        vector<float> angles= values->read_float_array();
        vector<float> speeds= values->read_float_array();
        vector<float> loads= values->read_float_array();
        vector<float> goals= values->read_float_array();

        for(int i=0; i < ids.size(); ++i)
        {
            if(idMotors.count(ids[i]))
            {
                idMotors[ids[i]]->setLastUpdate(time(NULL));

                idMotors[ids[i]]->setCurrentAngle((double)angles[i]);
                if(idMotors[ids[i]]->getGoalAngleInit() == 0)
                    idMotors[ids[i]]->setGoalAngle(idMotors[ids[i]]->getAngle());

                idMotors[ids[i]]->setCurrentGoal((double)goals[i]);

                idMotors[ids[i]]->setCurrentSpeed((double)speeds[i]/1023);
                if(idMotors[ids[i]]->getGoalSpeedInit() == 0)
                {
                    if(idMotors[ids[i]]->getSpeed() > 0)
                        idMotors[ids[i]]->setGoalSpeed(idMotors[ids[i]]->getSpeed());
                    else
                        idMotors[ids[i]]->setGoalSpeed(0);
                }

                idMotors[ids[i]]->setCurrentLoad((double)loads[i]/1023);
                if(idMotors[ids[i]]->getGoalLoadInit() == 0)
                {
                    if(idMotors[ids[i]]->getLoad() > 0)
                        idMotors[ids[i]]->setGoalLoad(idMotors[ids[i]]->getLoad());
                    else
                        idMotors[ids[i]]->setGoalLoad(0);
                }
            }
        }
    }

    void Motors::goToZero(int duration, bool verbose)
    {
        connection->SchedulerTakeRefPosition("Zero", 1000);
    }

    void Motors::goToInit(int duration, bool verbose)
    {
        try
        {
            connection->SchedulerTakeRefPosition("Ini", 1000);
        }
        catch(string exc)
        {
            throw string("Failed to go to initial posiiton:\n\t") + exc;
        }
    }

    void Motors::raiseLoad(int duration, bool verbose)
    {
        map<string, Motor *>::iterator it;
        int cs = duration * 100;

        for(int i=0; i < cs; i++)
        {
            if(verbose)
                printf("\rLoad: %3d%%", ((i+1)*100)/cs);

            for(it = motors.begin(); it != motors.end(); ++it)
                it->second->setLoad(((double)i+1)/(double)cs);

            syst_wait_ms(10);
            pushValues();
        }
        if(verbose)
            cout << "" << endl;
    }

    void Motors::execute()
    {
        while(running)
        {
            connection->ServosGetValues_callback(1, motorsValues, this);
            pushValues();
            syst_wait_ms(1000/frequency);
        }
    }

    void Motors::scan()
    {
        map<string, Motor *>::iterator it;
        for(it = motors.begin(); it != motors.end(); ++it)
            it->second->setLastUpdate(0);
        connection->ServosScan(250, "Normal");
    }

    void Motors::setConnection(Connection *connection)
    {
        this->connection = connection;
    }

    Connection * Motors::getConnection()
    {
        return connection;
    }

    void Motors::setConfiguration(MoveSchedulerConfig *config)
    {
        this->configuration = config;
    }

    MoveSchedulerConfig *Motors::getConfiguration()
    {
        return configuration;
    }

    void Motors::setMotors(map<string, Motor *> motors)
    {
        this->motors = motors;
    }

    map<string, Motor *> Motors::getMotors()
    {
        return motors;
    }

    void Motors::setIdMotors(map<int, Motor *> idMotors)
    {
        this->idMotors = idMotors;
    }

    map<int, Motor *> Motors::getIdMotors()
    {
        return idMotors;
    }

    void Motors::setRunning(bool val)
    {
        this->running = val;
    }

    bool Motors::getRunning()
    {
        return running;
    }

    void Motors::setFrequency(int frequency)
    {
        this->frequency = frequency;
    }

    int Motors::getFrequency()
    {
        return frequency;
    }
}
